Canvas: Dynamic connection direction

adela-almasan opened the pull request:
https://github.com/user-attachments/assets/6fad7e6d-4294-4002-a430-bb965abb70a9

To Do:

- [x] Direction editor
- [x] Update UI
- [x] Migration
- [x] Tests
- [x] Update gdev
- [x] Docs
- [ ] What's new



Fixes https://github.com/grafana/grafana/issues/92369

[test dashboard](https://github.com/user-attachments/files/21919875/debug-New.panel-2025-08-21.09_33_07.json.txt)


Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

adela-almasan committed:
DirectionDimension cue gen

adela-almasan committed:
DirectionDimensionEditor

adela-almasan committed:
add ConnectionDirection to schema

adela-almasan committed:
update ui

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
migration

adela-almasan commented on packages/grafana-schema/src/common/dimensions.cue:
```diff
@@ -44,6 +44,15 @@ ResourceDimensionMode: "fixed" | "field" | "mapping" @cuetsy(kind="enum")
+  fixed?: string
 }@cuetsy(kind="interface")
 
+ConnectionDirection: "forward" | "reverse" | "both" | "none" @cuetsy(kind="enum", memberNames="Forward|Reverse|Both|None")
```
This probably shouldn't be here, but not sure where's the best place for it

adela-almasan committed:
update docs

adela-almasan committed:
fix default direction

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
gdev update

adela-almasan committed:
docs update

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

# Conflicts:
#	docs/sources/panels-visualizations/visualizations/canvas/index.md

adela-almasan committed:
updated gdev version; Network Weathermap - updated direction to use fields

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
lint

drew08t commented:
> > Something isn't initializing correctly. When I first load `Panel Tests - Canvas Connection Examples` `Data Pipeline Flow Monitor` all connection directions, even those that are fixed, are not rendering arrow heads. It isn't until I change the options that arrow heads appear.
> > <img alt="image" width="2000" height="664" src="https://private-user-images.githubusercontent.com/60050885/471727828-d645a80d-d525-4eb4-b0db-894a27619a50.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3NDE2MzUsIm5iZiI6MTc1Mzc0MTMzNSwicGF0aCI6Ii82MDA1MDg4NS80NzE3Mjc4MjgtZDY0NWE4MGQtZDUyNS00ZWI0LWIwZGItODk0YTI3NjE5YTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI4VDIyMjIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwMWRhOTk0MTUxY2MxOTY4MGZhZGI4YjE4ZjE1MDBiZjNkZjgyODlkOTc2NzNjOTAzMDYxMDA4ZTEzMzc0ZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.emiyg59mE_moSoD-AUTKaaoJNoRHU7X1IRbDz3Y3QaU">
> 
> @drew08t does this happen with the other panels too? I might've messed up that panel, I actually thought I removed it..

Not that I can see. After making this change to one connection, it works during init.

<img width="787" height="458" alt="image" src="https://github.com/user-attachments/assets/e0392027-dbfe-4da8-88a0-9dd9dbedd83a" />

adela-almasan commented:
> > > Something isn't initializing correctly. When I first load `Panel Tests - Canvas Connection Examples` `Data Pipeline Flow Monitor` all connection directions, even those that are fixed, are not rendering arrow heads. It isn't until I change the options that arrow heads appear.
> > > <img alt="image" width="2000" height="664" src="https://private-user-images.githubusercontent.com/60050885/471727828-d645a80d-d525-4eb4-b0db-894a27619a50.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3NDE2MzUsIm5iZiI6MTc1Mzc0MTMzNSwicGF0aCI6Ii82MDA1MDg4NS80NzE3Mjc4MjgtZDY0NWE4MGQtZDUyNS00ZWI0LWIwZGItODk0YTI3NjE5YTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI4VDIyMjIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwMWRhOTk0MTUxY2MxOTY4MGZhZGI4YjE4ZjE1MDBiZjNkZjgyODlkOTc2NzNjOTAzMDYxMDA4ZTEzMzc0ZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.emiyg59mE_moSoD-AUTKaaoJNoRHU7X1IRbDz3Y3QaU">
> > 
> > 
> > @drew08t does this happen with the other panels too? I might've messed up that panel, I actually thought I removed it..
> 
> Not that I can see. After making this change to one connection, it works during init.
> 
> <img alt="image" width="787" height="458" src="https://private-user-images.githubusercontent.com/60050885/471733308-e0392027-dbfe-4da8-88a0-9dd9dbedd83a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3NDIwNDQsIm5iZiI6MTc1Mzc0MTc0NCwicGF0aCI6Ii82MDA1MDg4NS80NzE3MzMzMDgtZTAzOTIwMjctZGJmZS00ZGE4LTg4YTAtOWRkOWRiZWRkODNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI4VDIyMjkwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiOWY3NmIwMjZiMmExNmQ3ZDZmNmIxMTRkODM3MzYzNzI0MDcyNmRkMWY0NTViMmNkOTAzY2Q0MjZlYmU2MzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5ETQR_Yk6aA55F2Kx-J6P7OWS_U9J8yWFeA0-xlvAq4">

ooh yeah, the gdev is broken then, fixed is not an object. I'll remove the panel altogether, not sure how/when I added it back. I updated Network Weathermap panel to use fields for direction.

adela-almasan committed:
remove Data Pipeline Flow Monitor panel

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
fix migration

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

# Conflicts:
#	public/app/features/canvas/runtime/scene.tsx

adela-almasan committed:
migration++

Develer commented on public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:
Awesome PR, love it 😻
Could you also add those changes from ConnectionSVG to ConnectionSVG2? Or I can do it myself, just let me know 🙂

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
connection direction changes in ConnectionSVG2

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

# Conflicts:
#	devenv/dev-dashboards/panel-canvas/canvas-connection-examples.json

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

adela-almasan committed:
canvas shouldMigrate condition

adela-almasan committed:
shouldMigrate by version

fastfrwrd commented on public/app/features/dimensions/direction.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  const getDirectionFromValue = (value: unknown): ConnectionDirection => {
+    if (value === null || value === undefined) {
```
could do the more terse `value == null`

adela-almasan committed:
nulling

adela-almasan committed:
Merge branch 'main' into adela/connection_direction

fastfrwrd reviewed (approved):
Looks good!
